Add initial support for CSSStyleSheet constructor and TreeScope.adoptedStyleSheets attribute

cdumez opened the pull request:
#### 6f86bd6200eb4f6528757f1744f49fa653442882
<pre>
Add initial support for CSSStyleSheet constructor and TreeScope.adoptedStyleSheets attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=245648">https://bugs.webkit.org/show_bug.cgi?id=245648</a>

Reviewed by Antti Koivisto and Brent Fulgham.

Add initial support for CSSStyleSheet constructor and TreeScope.adoptedStyleSheets attribute:
- <a href="https://w3c.github.io/csswg-drafts/cssom-1/#dom-cssstylesheet-cssstylesheet">https://w3c.github.io/csswg-drafts/cssom-1/#dom-cssstylesheet-cssstylesheet</a>
- <a href="https://w3c.github.io/csswg-drafts/cssom-1/#extensions-to-the-document-or-shadow-root-interface">https://w3c.github.io/csswg-drafts/cssom-1/#extensions-to-the-document-or-shadow-root-interface</a>

The implementation should be complete and is passing most tests. One issue that remains is that
DocumentOrShadowRoot.adoptedStyleSheets should be an ObservableArray instead of a FrozenArray.
We don&apos;t support for ObservableArray yet so this will need a follow-up patch.

The new support is behind an experimental feature flag, currently off by default.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-baseURL.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-cssRules-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-disabled-regular-sheet-insertion-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-disallow-import.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-duplicate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-replace-on-regular-sheet-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-modify-after-removal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-template-adoption-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/adoptedstylesheets-observablearray-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/insertRule-across-context-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::create):
(WebCore::CSSStyleSheet::CSSStyleSheet):
(WebCore::CSSStyleSheet::didMutateRules):
(WebCore::CSSStyleSheet::didMutate):
(WebCore::CSSStyleSheet::forEachStyleScope):
(WebCore::CSSStyleSheet::setDisabled):
(WebCore::CSSStyleSheet::insertRule):
(WebCore::CSSStyleSheet::replace):
(WebCore::CSSStyleSheet::replaceSync):
(WebCore::CSSStyleSheet::constructorDocument const):
(WebCore::CSSStyleSheet::addAdoptingTreeScope):
(WebCore::CSSStyleSheet::removeAdoptingTreeScope):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSStyleSheet.idl:
* Source/WebCore/css/DocumentOrShadowRoot+CSSOM.idl:
* Source/WebCore/css/MediaList.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::parseString):
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSheet):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseStyleSheet):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::moveShadowTreeToNewDocument):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::insertedIntoAncestor):
(WebCore::ShadowRoot::moveShadowRootToNewParentScope):
(WebCore::ShadowRoot::moveShadowRootToNewDocument):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::~TreeScope):
(WebCore::TreeScope::adoptedStyleSheets const):
(WebCore::TreeScope::setAdoptedStyleSheets):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::collectActiveStyleSheets):
(WebCore::Style::Scope::treeScope):
* Source/WebCore/style/StyleScope.h:

Canonical link: <a href="https://commits.webkit.org/254894@main">https://commits.webkit.org/254894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9170f01399ab4839071c994e59b965ea1045152

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35149 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99897 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158252 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94575 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33649 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82929 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96330 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96220 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77410 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26626 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81561 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81361 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69648 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/children-changed, /TestWTF:WTF_WordLock.ManyContendedLongSections (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82140 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34748 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15398 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77086 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32563 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16376 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26618 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3421 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36328 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79672 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38242 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35465 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17459 "Failed to checkout and rebase branch from PR 4690") | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -519,6 +519,18 @@ CompressionStreamEnabled:
+ConstructableStylesheetObjectsEnabled:
+  type: bool
+  humanReadableName: "Constructable Stylesheet Objects API"
+  humanReadableDescription: "Enable Constructable Stylesheet Objects API API"
```
API API

anttijk commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -519,6 +519,18 @@ CompressionStreamEnabled:
 
+ConstructableStylesheetObjectsEnabled:
+  type: bool
+  humanReadableName: "Constructable Stylesheet Objects API"
```
I don't think "Objects" adds anything here, just Constructable Stylesheets is sufficient.

(seems like that was the title of a draft spec).

anttijk commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -165,36 +201,45 @@ void CSSStyleSheet::didMutateRules(RuleMutationType mutationType, WhetherContent
-        if (insertedKeyframesRule) {
-            if (auto* resolver = scope->resolverIfExists())
-                resolver->addKeyframeStyle(*insertedKeyframesRule);
+    forEachStyleScope([&](Style::Scope& scope) {
```
Interesting that a CSSStyleSheet can now have multiple clients, something that has been avoided so far. This should work, but I hope there are lots of WPTs about mutations.

brentfulgham commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -88,6 +91,21 @@ Ref<CSSStyleSheet> CSSStyleSheet::createInline(Ref<StyleSheetContents>&& sheet,
+    else {
+        baseURL = document.completeURL(init.baseURL);
+        if (!baseURL.isValid())
+            return Exception { NotAllowedError, "Base URL is invalid"_s };
```
I wonder what URL the test uses that doesn't throw here? (from earlier test output): 
"FAIL: Constructing sheet with invalid base URL throws a NotAllowedError"

cdumez replied:
The test is using "chrome://". It seems to be treated as valid by our implementation though. I am not sure if this is an issue with the test or our implementation. Also note that note of the browsers seems to be passing this test at the moment.

cdumez replied:
`u = new URL("chrome://");` throws in Chrome and Firefox, saying the URL is invalid. However, it doesn't throw in Safari so it seems the behavior difference is in our URL parser (cc @achristensen07). I know we're the only browser matching the URL spec in some cases so this may be one of these cases. We may need to update the test to use a URL that is recognized as invalid in all browsers.

brentfulgham commented on Source/WebCore/css/DocumentOrShadowRoot+CSSOM.idl:
```diff
@@ -26,5 +26,8 @@
 partial interface mixin DocumentOrShadowRoot {
     [SameObject] readonly attribute StyleSheetList styleSheets;
+
+    // FIXME: This should use an ObservableArray<CSSStyleSheet> but we don't support this yet.
```
Do we have a Bugzilla we could reference in this FIXME?

cdumez replied:
https://bugs.webkit.org/show_bug.cgi?id=238281 is the one about adding support for ObservableArray. I guess we need one about adopting ObservableArray here too.

cdumez replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=245689. Will update the comment.

brentfulgham commented on Source/WebCore/dom/TreeScope.cpp:
```diff
@@ -54,6 +55,7 @@ namespace WebCore {
 
 struct SameSizeAsTreeScope {
     void* pointers[10];
+    Vector<void*> vector;
```
This doesn't seem to be used yet?

cdumez replied:
Those are not used. SameSizeAsTreeScope is merely supposed to have the same size as TreeScope. Since I added a Vector to TreeScope, I added one here too so static assertions don't hit.

brentfulgham reviewed (approved):
LGTM.

anttijk commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -2855,7 +2855,6 @@ sub GenerateHeader
     my $interfaceName = $interface->type->name;
     my $className = "JS$interfaceName";
     my %structureFlags = ();
-
```
random change

cdumez replied:
I had to make this change to work around an incremental build issue with the GTK port.

anttijk commented on Source/WebCore/css/CSSStyleSheet.h:
```diff
@@ -139,6 +161,9 @@ class CSSStyleSheet final : public StyleSheet {
     CSSStyleSheet(Ref<StyleSheetContents>&&, Node& ownerNode, const TextPosition& startPosition, bool isInlineStylesheet, const std::optional<bool>&);
+    CSSStyleSheet(Ref<StyleSheetContents>&&, Document&, Init&&);
+
+    void forEachStyleScope(const Function<void(Style::Scope&)>&);
```
I think this is only ever used internally in this class? It would be better not to expose it as public. Also it could take template argument instead of Function and perhaps even be a standalone helper in cpp?

cdumez replied:
It is already private. Making it a standalone in the cpp is a bit annoying since it relies on CSSStyleSheet data members.

anttijk commented on Source/WebCore/css/parser/CSSParser.h:
```diff
@@ -63,7 +63,8 @@ class CSSParser {
 
-    void parseSheet(StyleSheetContents&, const String&);
+    enum class IgnoreImportRules : bool { No, Yes };
+    void parseSheet(StyleSheetContents&, const String&, IgnoreImportRules = IgnoreImportRules::No);
```
I think IgnoreImportRules could be a bit in CSSParserContext and you could avoid passing it around as argument.

cdumez committed:
Add initial support for CSSStyleSheet constructor and TreeScope.adoptedStyleSheets attribute
https://bugs.webkit.org/show_bug.cgi?id=245648

Reviewed by Antti Koivisto and Brent Fulgham.

Add initial support for CSSStyleSheet constructor and TreeScope.adoptedStyleSheets attribute:
- https://w3c.github.io/csswg-drafts/cssom-1/#dom-cssstylesheet-cssstylesheet
- https://w3c.github.io/csswg-drafts/cssom-1/#extensions-to-the-document-or-shadow-root-interface

The implementation should be complete and is passing most tests. One issue that remains is that
DocumentOrShadowRoot.adoptedStyleSheets should be an ObservableArray instead of a FrozenArray.
We don't support for ObservableArray yet so this will need a follow-up patch.

The new support is behind an experimental feature flag, currently off by default.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-baseURL.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-cssRules-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-disabled-regular-sheet-insertion-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-disallow-import.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-duplicate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-constructable-replace-on-regular-sheet-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-modify-after-removal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleSheet-template-adoption-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/adoptedstylesheets-observablearray-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/insertRule-across-context-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::create):
(WebCore::CSSStyleSheet::CSSStyleSheet):
(WebCore::CSSStyleSheet::didMutateRules):
(WebCore::CSSStyleSheet::didMutate):
(WebCore::CSSStyleSheet::forEachStyleScope):
(WebCore::CSSStyleSheet::setDisabled):
(WebCore::CSSStyleSheet::insertRule):
(WebCore::CSSStyleSheet::replace):
(WebCore::CSSStyleSheet::replaceSync):
(WebCore::CSSStyleSheet::constructorDocument const):
(WebCore::CSSStyleSheet::addAdoptingTreeScope):
(WebCore::CSSStyleSheet::removeAdoptingTreeScope):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSStyleSheet.idl:
* Source/WebCore/css/DocumentOrShadowRoot+CSSOM.idl:
* Source/WebCore/css/MediaList.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::parseString):
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSheet):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseStyleSheet):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::moveShadowTreeToNewDocument):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::insertedIntoAncestor):
(WebCore::ShadowRoot::moveShadowRootToNewParentScope):
(WebCore::ShadowRoot::moveShadowRootToNewDocument):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::~TreeScope):
(WebCore::TreeScope::adoptedStyleSheets const):
(WebCore::TreeScope::setAdoptedStyleSheets):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::collectActiveStyleSheets):
(WebCore::Style::Scope::treeScope):
* Source/WebCore/style/StyleScope.h:

Canonical link: https://commits.webkit.org/254894@main

anttijk commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -73,6 +73,7 @@ struct CSSParserContext {
     bool gradientPremultipliedAlphaInterpolationEnabled { false };
     bool gradientInterpolationColorSpacesEnabled { false };
     bool subgridEnabled { false };
+    bool shouldIgnoreImportRules { false };
```
This should be included in operator==.

While the class is not super well organised it would still be good to put the bit further up. This section is for enable bits coming from settings.

anttijk commented on Source/WebCore/css/MediaList.h:
```diff
@@ -96,6 +96,7 @@ class MediaList final : public RefCounted<MediaList> {
     void clearParentStyleSheet() { ASSERT(m_parentStyleSheet); m_parentStyleSheet = nullptr; }
     void clearParentRule() { ASSERT(m_parentRule); m_parentRule = nullptr; }
     const MediaQuerySet* queries() const { return m_mediaQueries.get(); }
+    MediaQuerySet* queries() { return m_mediaQueries.get(); }
```
Where is this needed?

cdumez replied:
I needed it for this logic inside the CSSStyleSheet constructor:
```
    WTF::switchOn(WTFMove(options.media), [this](RefPtr<MediaList>&& mediaList) {
         if (auto* queries = mediaList->queries())
             setMediaQueries(*queries);
```

anttijk commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -189,9 +189,11 @@ RefPtr<StyleRuleBase> CSSParserImpl::parseRule(const String& string, const CSSPa
 {
     CSSParserImpl parser(context, string, &styleSheet, nullptr);
-    bool firstRuleValid = parser.consumeRuleList(parser.tokenizer()->tokenRange(), TopLevelRuleList, [&styleSheet](RefPtr<StyleRuleBase> rule) {
+    bool firstRuleValid = parser.consumeRuleList(parser.tokenizer()->tokenRange(), TopLevelRuleList, [&styleSheet, shouldIgnoreImportRules = context.shouldIgnoreImportRules](RefPtr<StyleRuleBase> rule) {
```
I'd just use [&] capture

anttijk commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2041,7 +2041,7 @@ inline void Node::moveShadowTreeToNewDocument(ShadowRoot& shadowRoot, Document&
         node.moveNodeToNewDocument(oldDocument, newDocument);
     }, [&oldDocument, &newDocument](ShadowRoot& innerShadowRoot) {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(&innerShadowRoot.document() == &oldDocument);
-        innerShadowRoot.moveShadowRootToNewDocument(newDocument);
+        innerShadowRoot.moveShadowRootToNewDocument(oldDocument, newDocument);
```
This new parameter shouldn't be needed. Per assertion above you can just read ShadowRoot::document().

cdumez replied:
There are actually 2 call sites. For this one, I would agree. The issue is with the second call site:
```
void ShadowRoot::moveShadowRootToNewParentScope(TreeScope& newScope, Document& newDocument)
{
    auto& oldDocument = documentScope();
    setParentTreeScope(newScope);
    moveShadowRootToNewDocument(oldDocument, newDocument);
}
```

`setParentTreeScope(newScope)` updates the shadowRoot's documentScope to be newDocument, *before* moveShadowRootToNewDocument() gets called.
